Lib/subprocess.py

```diff
@@ -1480,7 +1480,21 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             if shell:
                 startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
                 startupinfo.wShowWindow = _winapi.SW_HIDE
-                comspec = os.environ.get("COMSPEC", "cmd.exe")
+                if not executable:
+                    # gh-101283: without a fully-qualified path, before Windows
+                    # checks the system directories, it first looks in the
+                    # application directory, and also the current directory if
+                    # NeedCurrentDirectoryForExePathW(ExeName) is true, so try
+                    # to avoid executing unqualified "cmd.exe".
+                    comspec = os.environ.get('ComSpec')
+                    if not comspec:
+                        system_root = os.environ.get('SystemRoot', '')
+                        comspec = os.path.join(system_root, 'System32', 'cmd.exe')
+                        if not os.path.isabs(comspec) or not os.path.isfile(comspec):
+                            raise FileNotFoundError('cmd.exe not found; neither %ComSpec% nor %SystemRoot% is set')
```

Comment:
`CreateProcessW()` will fail with `FileNotFoundError` if the shell path doesn't exist. It doesn't need to be checked twice, but we do need to check for a relative path. 

```suggestion
                        if not os.path.isabs(comspec):
                            raise FileNotFoundError(
                'shell not found: neither %ComSpec% nor %SystemRoot% is set')
```

Note in terms of security that `os.path.isabs()` is flawed on Windows. It's true for a relative rooted path such as "\\Windows", which depends on the drive or UNC share of the current working directory. However, it's unlikely that `SystemRoot` is set to a relative path. Generally it's either correct or missing.